independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/EncodedMediaType.java

```diff
@@ -1,20 +1,44 @@
 package org.jboss.resteasy.reactive.server.core;
 
+import java.nio.charset.StandardCharsets;
+import java.util.Objects;
 import javax.ws.rs.core.MediaType;
 import org.jboss.resteasy.reactive.server.spi.ContentType;
 
 /**
  * Wrapper around MediaType that saves the toString value, to avoid
  * the expensive header delegate processing.
+ * It also harmonizes the use of charset
  */
 public class EncodedMediaType implements ContentType {
     final MediaType mediaType;
+    final String charset;
     String encoded;
-    String charset;
 
     public EncodedMediaType(MediaType mediaType) {
-        this.mediaType = mediaType;
-        this.charset = mediaType.getParameters().get("charset");
+        MediaType effectiveMediaType = mediaType;
+        String effectiveCharset;
+        String originalCharset = mediaType.getParameters().get("charset");
+        if (isStringMediaType(mediaType)) {
+            effectiveCharset = originalCharset;
+            if (effectiveCharset == null) {
+                effectiveCharset = StandardCharsets.UTF_8.name();
+            }
+        } else {
+            // it doesn't make sense to add charset to non string types
+            effectiveCharset = null;
+        }
+        this.charset = effectiveCharset;
+        if (!Objects.equals(originalCharset, effectiveCharset)) {
+            effectiveMediaType = mediaType.withCharset(effectiveCharset);
+        }
+        this.mediaType = effectiveMediaType;
+    }
+
+    // TODO: does this need to be more complex?
+    private boolean isStringMediaType(MediaType mediaType) {
+        return mediaType.getType().equals("application")
```

Comment:
All `application/*` types require a charset? That doesn't seem right.

Reply:
Which ones should in your opinion. JSON and XML?

Reply:
Updated to be much more strict